src/pixie/fileformats/qoi.nim

```diff
@@ -15,158 +14,150 @@ const
-      dec(run)
+      dec run
     else:
       let b0 = data.readUint8(p)
-      inc(p)
-      case b0
+      inc p
+
+      case b0:
       of opRgb:
-        px.r = data.readUint8(p+0)
-        px.g = data.readUint8(p+1)
-        px.b = data.readUint8(p+2)
-        inc(p, 3)
+        px.r = data.readUint8(p + 0)
+        px.g = data.readUint8(p + 1)
+        px.b = data.readUint8(p + 2)
+        p += 3
       of opRgba:
-        px.r = data.readUint8(p+0)
-        px.g = data.readUint8(p+1)
-        px.b = data.readUint8(p+2)
-        px.a = data.readUint8(p+3)
-        inc(p, 4)
+        px.r = data.readUint8(p + 0)
+        px.g = data.readUint8(p + 1)
+        px.b = data.readUint8(p + 2)
+        px.a = data.readUint8(p + 3)
+        p += 4
       else:
-        case b0 and opMask2
+        case b0 and opMask2:
         of opIndex:
           px = index[b0]
         of opDiff:
-          px.r = px.r + uint8((b0 shr 4) and 0x03) - 2
-          px.g = px.g + uint8((b0 shr 2) and 0x03) - 2
-          px.b = px.b + uint8((b0 shr 0) and 0x03) - 2
+          px.r = px.r + ((b0 shr 4) and 0x03).uint8 - 2
+          px.g = px.g + ((b0 shr 2) and 0x03).uint8 - 2
+          px.b = px.b + ((b0 shr 0) and 0x03).uint8 - 2
         of opLuma:
-          let b1 = data.readUint8(p)
-          inc(p)
-          let vg = (b0.uint8 and 0x3f) - 32
+          let
+            b1 = data.readUint8(p)
+            vg = (b0.uint8 and 0x3f) - 32
           px.r = px.r + vg - 8 + ((b1 shr 4) and 0x0f)
           px.g = px.g + vg
           px.b = px.b + vg - 8 + ((b1 shr 0) and 0x0f)
+          inc p
         of opRun:
           run = b0 and 0x3f
-        else: assert false
+        else:
+          raise newException(PixieError, "Unexpected QOI op")
+
       index[hash(px)] = px
+
     dst = px
+
   while p < data.len:
-    case data[p]
-    of '\0': discard
-    of '\1': break # ignore trailing data
+    case data[p]:
+    of '\0':
+      discard
+    of '\1':
+      break # ignore trailing data
     else:
       raise newException(PixieError, "Invalid QOI padding")
     inc(p)
 
 proc decodeQoi*(data: string): Image {.raises: [PixieError].} =
   ## Decodes data in the QOI file format to an `Image`.
-  decompressQoi(data).toImage()
-
-proc decodeQoi*(data: seq[uint8]): Image {.inline, raises: [PixieError].} =
-  ## Decodes data in the QOI file format to an `Image`.
-  decodeQoi(cast[string](data))
+  newImage(decodeQoiRaw(data))
 
-proc compressQoi*(qoi: Qoi): string =
+proc encodeQoi*(qoi: Qoi): string {.raises: [PixieError].} =
   ## Encodes raw QOI pixels to the QOI file format.
+
+  if qoi.width.int * qoi.height.int > uint32.high.int:
+    raise newException(PixieError, "QOI is too large to encode")
+
+  # Allocate a buffer 3/4 the size of the pathological encoding
   result = newStringOfCap(14 + 8 + qoi.data.len * 3)
-    # allocate a buffer 3/4 the size of the pathological encoding
+
   result.add(qoiSignature)
-  when cpuEndian == bigEndian:
```

Comment:
no tests, untested. we do not support big endian anywhere else and have not / cannot? ensure this actually is all that is needed. rm.